src/haz3lcore/zipper/action/Perform.re

```diff
@@ -30,6 +30,32 @@ let go_z =
   module Move = Move.Make(M);
   module Select = Select.Make(M);
 
+  let refine = (z: Zipper.t) => {
+    switch (Indicated.ci_of(z, meta.statics.info_map)) {
+    | None => None
+    | Some(
+        InfoExp({
+          cls: Exp(EmptyHole),
+          status: NotInHole(Common(Ana(Consistent({ana, _})))),
+          ctx,
+          _,
+        }),
+      ) =>
+      let z =
+        Printer.zipper_of_string(
+          ~zipper_init=z,
+          switch (Typ.term_of(Typ.weak_head_normalize(ctx, ana))) {
+          | Arrow(_, _) => "fun -> "
+          | Prod(tys) =>
+            "(" ++ StringUtil.repeat(List.length(tys) - 1, ", ") ++ ") "
```

Comment:
This ends up adding a trailing space but that also seems to happen in the paste case as well.

Reply:
We don't want extra parens if we're inside of a function application but I don't know how much it matters.